File: script/utils/cron.lua
local _M = {}

function _M:add(expression, func)
  local c = cron(expression)
  local sched = {}
  sched.cron = c
  local run_func = function()
    if sched.timestamp == nil then
      local timestamp = sched.cron:next()
      sched.timestamp = timestamp
    end
    -- print(sched.timestamp)
    -- print(os.time())
    if sched.timestamp <= os.time() then
      func()
      sched.timestamp = sched.cron:next()
    end
  end
  sched.func = run_func
  if self._sched == nil then
    self._sched = {}
  end
  table.insert(self._sched, sched)
end

function _M:sched()
  return self._sched
end

return _M

File: src/cron.rs
use std::str::FromStr;

use chrono::Local;
use cron::{OwnedScheduleIterator, Schedule};
use mlua::prelude::*;

pub struct Cron(OwnedScheduleIterator<Local>);

pub fn create_cron(lua: &Lua) -> LuaResult<LuaFunction> {
    lua.create_function(|_, expression: String| {
        let schedule = Schedule::from_str(&expression).to_lua_err()?;
        let scheduler = schedule.upcoming_owned(Local);
        Ok(Cron(scheduler))
    })
}

impl LuaUserData for Cron {
    fn add_methods<'lua, M: LuaUserDataMethods<'lua, Self>>(_methods: &mut M) {
        _methods.add_method_mut("next", |_, this, ()| {
            let datetime = this.0.next();
            if let Some(datetime) = datetime {
                Ok(datetime.timestamp())
            } else {
                Ok(0)
            }
        });
    }
}
